src/main/scala/is/hail/expr/ir/ArraySorter.scala

```diff
@@ -8,27 +8,37 @@ import is.hail.utils._
   val typ: Type = array.elt
   val ti: TypeInfo[_] = typeToTypeInfo(typ)
-  val sortmb: EmitMethodBuilder = mb.fb.newMethod[Region, Int, Int, Unit]
+  val sortmb: EmitMethodBuilder = mb.fb.newMethod[Region, Int, Int, Boolean, Unit]
 
-  val (compare: CodeOrdering.F[Int], equiv: CodeOrdering.F[Boolean]) = if (keyOnly) {
+  val (less: CodeOrdering.F[Boolean], greater: CodeOrdering.F[Boolean], equiv: CodeOrdering.F[Boolean]) = if (keyOnly) {
     val ttype = coerce[TBaseStruct](typ)
     require(ttype.size == 2)
     val kt = ttype.types(0)
+    val mk1s = sortmb.newLocal[Boolean]
+    val mk2s = sortmb.newLocal[Boolean]
+
+    val clt: CodeOrdering.F[Boolean] = {
+      case (r1: Code[Region], (m1: Code[Boolean] @unchecked, v1: Code[Long] @unchecked),
+            r2: Code[Region], (m2: Code[Boolean] @unchecked, v2: Code[Long] @unchecked)) =>
+        val mk1 = Code(mk1s := ttype.isFieldMissing(r1, v1, 0), mk1s)
+        val mk2 = Code(mk2s := ttype.isFieldMissing(r2, v2, 0), mk2s)
+        val k1 = mk1s.mux(defaultValue(kt), r1.loadIRIntermediate(kt)(ttype.fieldOffset(v1, 0)))
+        val k2 = mk2s.mux(defaultValue(kt), r2.loadIRIntermediate(kt)(ttype.fieldOffset(v2, 0)))
+        val cmp = mb.getCodeOrdering[Int](kt, CodeOrdering.compare, missingGreatest = true)(r1, (mk1, k1), r2, (mk2, k2))
+        !m1 && (m2 || cmp < 0)
+    }
 
-    val comp: CodeOrdering.F[Int] = {
-      val mk1l = sortmb.newLocal[Boolean]
-      val mk2l = sortmb.newLocal[Boolean]
-
-      { case (r1: Code[Region], (m1: Code[Boolean]@unchecked, v1: Code[Long]@unchecked),
-              r2: Code[Region], (m2: Code[Boolean]@unchecked, v2: Code[Long]@unchecked)) =>
-          val mk1 = Code(mk1l := ttype.isFieldMissing(r1, v1, 0), mk1l)
-          val mk2 = Code(mk2l := ttype.isFieldMissing(r2, v2, 0), mk2l)
-          val k1 = mk1l.mux(defaultValue(kt), r1.loadIRIntermediate(kt)(ttype.fieldOffset(v1, 0)))
-          val k2 = mk2l.mux(defaultValue(kt), r2.loadIRIntermediate(kt)(ttype.fieldOffset(v2, 0)))
-          val cmp = mb.getCodeOrdering[Int](kt, CodeOrdering.compare, missingGreatest = true)(r1, (mk1, k1), r2, (mk2, k2))
-          (m1 || m2).mux((m1 && m2).mux(0, m1.mux(1, -1)), cmp)
-      }
+    val cgt: CodeOrdering.F[Boolean] = {
+      case (r1: Code[Region], (m1: Code[Boolean] @unchecked, v1: Code[Long] @unchecked),
+            r2: Code[Region], (m2: Code[Boolean] @unchecked, v2: Code[Long] @unchecked)) =>
+        val mk1 = Code(mk1s := ttype.isFieldMissing(r1, v1, 0), mk1s)
+        val mk2 = Code(mk2s := ttype.isFieldMissing(r2, v2, 0), mk2s)
+        val k1 = mk1s.mux(defaultValue(kt), r1.loadIRIntermediate(kt)(ttype.fieldOffset(v1, 0)))
+        val k2 = mk2s.mux(defaultValue(kt), r2.loadIRIntermediate(kt)(ttype.fieldOffset(v2, 0)))
+        val cmp = mb.getCodeOrdering[Int](kt, CodeOrdering.compare, missingGreatest = false)(r1, (mk1, k1), r2, (mk2, k2))
+        !m1 && (m2 || cmp > 0)
```

Comment:
Here I split the ordering into two, depending on whether the sort is ascending or descending. Missing values will come at the end in either case.

Before, the comparison function was only used as `comp < 0`, so I simplified it to a less-than function.